### ISSUE TITLE
[Reactor Gen4] Ensuring Reactor 3 can run along Reactor 4

### ISSUE THREAD

**Issue by simonbasle** [reactor/reactor#704]
### Update: this has been postponed, see announcement [in issue #705](https://github.com/reactor/reactor/issues/705#issuecomment-1025976281)
------

In order for that to be possible, several conditions:

 - [x] Netty 5 must be able to run alongside Netty 4 (confirmed by the Netty team, this is planned even for natives)
 - [ ] The package structure must be changed
 - [ ] The groupIds and/or artifactIds must be changed

## Choosing a new package structure

NB: This also gives us an opportunity to clear the package structure for enabling of modularization.

Current package root is `reactor`:

```java
reactor
-|---- reactor-core repo
 |_ core
 |_ adapter /!\ duplicated in addons (but actually just the reactor.adapter package)
 |_ util 
 |_ test
 |_ tools
 | |_ agent
-|----- reactor-addons repo
 |_ extra
 | |_ processor
 |------------ these are problematic
 |_ bool
 |_ cache
 |_ function
 |_ math
 |_ retry
 |_ scheduler
 |_ swing
 |_ adapter  /!\ duplicated in core (but not an actual package)
   |_ akka (the reactor.adapter.akka package)
   |_ rxjava (the reactor.adapter.rxjava package)
-|----- reactor-pool repo
 |_ pool
-|----- reactor-netty repo
 |_ netty
 | |_ various modules
-|----- kafka/kotlin/rabbitmq
 |_ kafka
 | |_ receiver
 | |_ sender
 | |_ tools
 | |_ samples
 |_ kotlin
 |_ rabbitmq //core classes here at root, the reactor.rabbitmq package
 | |_ samples // the reactor.rabbitmq.samples package
-|----- reactor-scala-extensions repo
 |_ core /!\ not ideal, but not an actual package either
 | |_ scala the reactor.core.scala package
 |_ util /!\ not ideal, but not an actual package either
 | |_ scala the reactor.util.scala package
```

Better view it as a list of packages:

```
reactor.core.*[!scala]
reactor.adapter
reactor.util.*[!scala]
reactor.test.*
reactor.tools.agent.*

//= addons
reactor.extra.processor
//------------ these are one level too low
reactor.bool
reactor.cache
reactor.function
reactor.math
reactor.retry
reactor.scheduler
reactor.swing
reactor.adapter.akka
reactor.adapter.rxjava

//= pool
reactor.pool.*

//= netty
reactor.netty.*

//= kafka/kotlin/rabbitmq
reactor.kafka.receiver
reactor.kafka.sender
reactor.kafka.tools
reactor.kafka.samples

reactor.kotlin.core.publisher
reactor.kotlin.core.util.function
reactor.kotlin.adapter.rxjava
reactor.kotlin.extra.*
reactor.kotlin.test

reactor.rabbitmq
reactor.rabbitmq.samples

//= reactor-scala-extensions repo
reactor.core.scala
reactor.util.scala
```

### Option 1: use a codename

The root would be changed to `reactor.codename`.

Using a codename is a suggestion by Sam following the experience of JUnit5.
The rationale is that adding a version number in the package could lead to confusing outcome if there is a gen5 release
without enough changes that another package restructuring is made.

For example OkHttp v4 still has okhttp3 in its package structure. (that's because they aimed at 100% backward compatibility when they switched to kotlin as dev language).

That being said, that probably assumes semantic versioning. If we do a Gen5, won't that be as breaking as this one?

The other difficulty is in finding a codename, which could be confused with BOM codenames.
A list of (single-word) things that can be used for a future generation (so the list isn't necessarily long).

#### Gemstones ?
 - starting at D, 4th letter of the alphabet
 - use well known gems alphabetically, but allow skipping letters
 - Diamond
 - Emerald
 - Jade
 - Lazurite

:warning: ❓Too close to elements codenames ?

See bigger list at https://en.wikipedia.org/wiki/List_of_gemstones_by_species

Examples:

 ```
 reactor.diamond.core.publisher
 reactor.diamond.util
 reactor.diamond.netty.http
 ```

### Power of 10 SI prefixes

See list at https://en.wikipedia.org/wiki/Metric_prefix

 - skipping 3 first deca,hecto,kilo
 - mega
 - giga
 - tera
 - peta
 - exa

My currently preferred option.

Examples:

```
reactor.mega.core.publisher
reactor.mega.util
reactor.mega.netty.http
```

### Option 2: simply align with the site => `io/projectreactor`

This has the drawback of not being future proof: if we ever want to do a Gen5, we will have to figure out a meaningful new root
(IF Gen4 and Gen5 need to coexist, which they'll probably do).

Examples:

```
io.projectreactor.core.publisher
io.projectreactor.util
io.projectreactor.netty.http
```

### Option 3: `io.projectreactor.reactor4`

Examples:

```
io.projectreactor.reactor4.core.publisher
io.projectreactor.reactor4.util
io.projectreactor.reactor4.netty.http
```

There have been comments (eg. [here](https://github.com/reactor/reactor/issues/703#issuecomment-931463719)) that having a number in there might cause issues with modularization?

### Suggested new package structure (module compatible)

No two modules can export the same packages.
But there is no such thing as actual "subpackages", so module 1 can export com.foo while module 2 still can export com.foo.bar.

Repackaging and cleaning package strucuture mainly implies the following changes:
 - [ ] use whichever `newReactorRoot` we decide
 - [ ] move flow adapter to reactor-adapters as adapter.jdk ?
 - [ ] move all reactor-extra packages under the `newReactorRoot/extra` root
 - [ ] change the package of scala extensions to have `scala` first then `core`/`util`

```
reactor.mega.core.*
//~adapter moved~
reactor.mega.util.*
reactor.mega.test.*
reactor.mega.tools.agent.*

//= addons
reactor.mega.extra.processor
reactor.mega.extra.bool
reactor.mega.extra.cache
reactor.mega.extra.function
reactor.mega.extra.math
reactor.mega.extra.retry
reactor.mega.extra.scheduler
reactor.mega.extra.swing
reactor.mega.adapter.akka
reactor.mega.adapter.rxjava
reactor.mega.adapter.jdk //from core

//= pool
reactor.mega.pool.*

//= netty
reactor.mega.netty.*

//= kafka/kotlin/rabbitmq
reactor.mega.kafka.receiver
reactor.mega.kafka.sender
reactor.mega.kafka.tools
reactor.mega.kafka.samples

reactor.mega.kotlin.core.publisher
reactor.mega.kotlin.core.util.function
reactor.mega.kotlin.adapter.rxjava
reactor.mega.kotlin.extra.bool
reactor.mega.kotlin.extra.math
reactor.mega.kotlin.extra.retry
reactor.mega.kotlin.extra.swing
reactor.mega.kotlin.test

reactor.mega.rabbitmq
reactor.mega.rabbitmq.samples

//= reactor-scala-extensions repo
reactor.mega.scala.core
reactor.mega.scala.util
```

**Comment by msgilligan** [comment]
I would vote for `io.projectreactor.mega.core.*`

**Comment by bric3** [comment]
As a user I always found the release train difficult to grasp compared to semantic versions. For this release 4, going option 2 is ok. But having the version in the package allows to easily identify the version at hand.

**Comment by simonbasle** [comment]
### Maven coordinates issue with option 2

reviving the discussion since there is one aspect that I haven't taken into account: the Maven coordinates / groupId :(

in option 2, we'd have a package root that matches the **current** groupId. But in order to allow mixing version, we need to change that groupId...

one might say we do have a difference today between the package root and the groupId, but that's not terribly clean and the repackaging is an opportunity to make things better. furthermore, I think the difference between the two was big enough that it was _slightly_ less confusing. If we have something close but not perfectly aligned (at least for reactor-core), it would further the confusion.

in that light, maybe @msgilligan suggestion above (option 3 + codename) makes sense.

### Number in the groupId / package
Per the above we probably want to align the "root package" with the `groupId` (or at least the root of the groupId).

@bric3 in order to support good modularity, it seems we would need to avoid having numbers in packages, as per @msgilligan [comment](https://github.com/reactor/reactor/issues/703#issuecomment-931463719) in the original issue.
 Codenames have also been recommended by @sbrannen with the experience and insights of Junit5 / JUnit Jupiter.

### Option 4: reverse-namespace + codename: `io.projectreactor.mega`
If we consider this, a 4th option surfaces: reverse namespace + codename.

Going with the SI prefixes, this would be `io.projectreactor.mega` root groupId and package.

If we expand a bit to a fuller sample of the various groupId and package structures:

<table>
<tr>
	<th>groupId
	<th>artifactId
	<th>package/folder structure
	<th>notes
<tr>
	<td>io.projectreactor.mega
	<td>reactor-core
	<td>io/projectreactor/mega/core/*
   <td>we could also use the raw root prefix of `io.projectreactor.mega` for the package structure, but I feel adding `core/` is a good thing. plus it clarifies this is indeed the true core of reactor ecosystem. compare to the netty case below.
<tr>
	<td>io.projectreactor.mega
	<td>reactor-test
	<td>io/projectreactor/mega/test/*
<tr>
	<td>io.projectreactor.mega
	<td>reactor-tools
	<td>io/projectreactor/mega/tools/*

<tr>
	<td>io.projectreactor.mega.addons
	<td>reactor-adapter
	<td>io/projectreactor/mega/addons/adapter/*
<tr>
	<td>io.projectreactor.mega.addons
	<td>reactor-extra
	<td>io/projectreactor/mega/addons/extra/*
<tr>
	<td>io.projectreactor.mega.addons
	<td>reactor-pool
	<td>io/projectreactor/mega/addons/pool/*

<tr>
	<td>io.projectreactor.mega.netty
	<td>reactor-netty-core
	<td>io/projectreactor/mega/netty/*
    <td>excluding modules like reactor-netty-http... we could perhaps consider adding `core/` to the package structure but this is probably not necessary to distinguish between the netty and non-netty modules.
<tr>
	<td>io.projectreactor.mega.netty
	<td>reactor-netty-http
	<td>io/projectreactor/mega/netty/http/*

<tr>
	<td>...
	<td>...
	<td>...

<tr>
	<td>io.projectreactor.mega.rabbitmq
	<td>reactor-rabbitmq
	<td>io/projectreactor/mega/rabbitmq/*
	<td>excluding the `samples` module
<tr>
	<td>io.projectreactor.mega.rabbitmq
	<td>reactor-rabbitmq-samples
	<td>io/projectreactor/mega/rabbitmq/samples/*

</table>

**Comment by simonbasle** [comment]
Note that unlike the release train codenames, the codename in the package/groupId would only reflect a foundational change like the gap between Reactor 3 and Reactor 4:

<table>
	<tr>
		<th>level
		<th>format
		<th>type of change
		<th>notes
	<tr>
		<td>package / groupId
		<td>io.projectreactor.codename
		<td>generational change
		<td>complete re-architecture, or change of JDK baseline. implies a need for running older generations in parallel
	<tr>
		<td>BOM version, minus the service release number
		<td>`year_of_inception.differentiator`.0 (eg. `2020.1`.0)
		<td>"cycle", allowing breaking changes to some extent
		<td>not strictly semantic versioning, but an intermediate level of migration effort (deprecations removal, reasonable amount of API changes...). there is an associated codename (eg. Europium) for internal discussions. the .0 at then end (aka the service release number) is then increased with each patch release, see below.
	<tr>
		<td>artifact versions
		<td>MAJOR.MINOR.PATCH
		<td>maintenance releases
		<td>not strictly semantic versioning since we can add APIs (but not remove: binary compatibility and api compatibility is maintained)
</table>
 



Even considering we don't strictly follow semantic versioning, there could be a reactor-core 4.1.0 and even a 5.0.0 version which could still be part of the `mega` 4th generation, but switching to `giga` 5th generation would 

There could still be

**Comment by simonbasle** [comment]
@msgilligan can you expand on your comment and give sources / explain what might go wrong:

> Please don't end the package (module) name with a number (...) because it will result in warnings during the build.

**Comment by msgilligan** [comment]
> @msgilligan can you expand on your comment and give sources / explain what might go wrong:

Nothing really goes wrong. It's just that the designers of the Java Module System wanted to discourage use of "version numbers" in module names, so they added an annoying warning in `javac` that looks like this:

`module name component rxjava3 should avoid terminal digits`

The complete warning looks like this:

```
1 warning
/Users/xxx/yyyy/tmpjars/io.reactivex.rxjava3/module-info.java:2: warning: [module] module name component rxjava3 should avoid terminal digits
open module io.reactivex.rxjava3 {
                        ^
```

**Comment by msgilligan** [comment]
p.s. A little bit of Googling found this: https://openjdk.java.net/projects/jigsaw/spec/issues/#VersionsInModuleNames

**Comment by bric3** [comment]
Bold style is from me.

> Resolution Abandon the previous proposal to mandate that module names appearing in source-form module declarations must both start and end with “Java letters”. **Revise the automatic-module naming algorithm to allow digits at the end of module names**. [EG discussion; proposal; discussion: Robert Scholte, Mark Reinhold, Tim Ellison; original proposal]

So numbers are allowed at the end.

**Comment by msgilligan** [comment]
> So numbers are allowed at the end.
 
Technically allowed, yes. But every (modular) user will get a warning message on every build. This should really be avoided.

**Comment by bric3** [comment]
Indeed warning message will be annoying. From this I don't know however if this warning is bound to be removed at some point.

**Comment by kasobol-msft** [comment]
Couple of questions about Reactor 3 and 4 interop.
Assuming that somebody ends up in a situation that they have reactive streams that are mixed, for example they're using a library that creates reactive stream with Reactor 3 and returns Mono or Flux from Reactor 3 but in their application they adopted Reactor 4 and they want to include library call in their Reactor 4 based chain.
- How would one consume Rector 3 Mono/Flux in Reactor 4? Reactive Streams interfaces would allow that, but is there anything more sophisticated planned needed (e.g. if I attempt to adapt v3 Flux to v4 Mono I'd like to catch that at compile time).
- Are there any plans to make v4 operators to be able to fuse with v3 operators if such opportunity exists?

**Comment by simonbasle** [comment]
Gen4 has been abandoned in favor of continuing the Gen3 with 3.5, see announcement [in issue #705](https://github.com/reactor/reactor/issues/705#issuecomment-1025976281)
------
